sdk/include/opentelemetry/sdk/metrics/state/async_metric_storage.h

```diff
@@ -48,21 +48,31 @@ class AsyncMetricStorage : public MetricStorage, public AsyncWritableMetricStora
     {
       auto aggr = DefaultAggregation::CreateAggregation(aggregation_type_, instrument_descriptor_);
       aggr->Aggregate(measurement.second);
+      cumulative_hashmap_lock_.lock();
       auto prev = cumulative_hash_map_->Get(measurement.first);
+      cumulative_hashmap_lock_.unlock();
       if (prev)
       {
         auto delta = prev->Diff(*aggr);
```

Comment:
> Here: auto delta = prev->Diff(*aggr); is executed without the lock.
> 
> Can two threads execute cumulative_hash_map_->Get() then both see prev, and both execute cumulative_hash_map_->Set(..., *delta)
> 
> It looks like a race, please clarify.
> 
> Maybe it would be better to only take the cumulative_hashmap_lock_ once, and execute:
> 
> get() if (prev) set(*delta) else set(*aggr)
> 
> within the same lock/unlock section.
> 
> If not needed for correctness, it still can help for performances (scalability), to avoid taking the lock twice.

Good point. I was trying to minimize the critical section with multiple locks, but it can add the race condition in future. Have simplified a bit by having a single lock for Record() method. Let me know if it doesn't make sense :)